mdsuite/file_io/extxyz_files.py

```diff
@@ -0,0 +1,305 @@
+        self.file_path = pathlib.Path(file_path).resolve()
+        self.n_header_lines = 2
+        if custom_data_map is None:
+            custom_data_map = {}
+        self.custom_data_map = custom_data_map
+
+        self._n_particles = None
+        self._mdata = None
+        self._properties_dict = None
+        self._species_dict = None
+        self._batch_size = None
+
+    def __str__(self):
+        return str(self.file_path)
+
+    def get_metadata(self):
+        with open(self.file_path, "r") as file:
+            # first header line: number of particles
+            self._n_particles = int(file.readline())
+            # second header line: properties
+            header = file.readline()
+
+            box_l = _get_box_l(header)
+
+            var_names_updated = copy.deepcopy(var_names)
+            var_names_updated.update(self.custom_data_map)
+            species_idx, property_dict = _get_property_summary(
+                header, var_names_updated
+            )
+            self._properties_dict = property_dict
+
+            file.seek(0)
+            self._species_dict = self._get_species_information(file, species_idx)
+
+            file.seek(0)
+            mdsuite.file_io.file_read.skip_n_lines(
+                file, self._n_particles + self.n_header_lines + 1
+            )
+            header_1 = file.readline()
+            sample_rate = int(round(_get_time(header_1) - _get_time(header)))
+
+            # get number of configs from file length
+            file.seek(0)
+            num_lines = sum(1 for _ in file)
+            n_configs_float = num_lines / (self._n_particles + self.n_header_lines)
+            n_configs = int(round(n_configs_float))
+            assert abs(n_configs_float - n_configs) < 1e-10
+
+        # same properties for all species
+        properties_list = []
+        for prop_name, prop_col_idxs in self._properties_dict.items():
+            properties_list.append(
+                mdsuite.database.simulation_database.PropertyInfo(
+                    name=prop_name, n_dims=len(prop_col_idxs)
+                )
+            )
+        species_list = []
+        for sp_name, sp_indices_dict in self._species_dict.items():
+            species_list.append(
+                mdsuite.database.simulation_database.SpeciesInfo(
+                    name=sp_name,
+                    n_particles=len(sp_indices_dict["line_idxs"]),
+                    properties=properties_list,
+                )
+            )
+
+        self._mdata = mdsuite.database.simulation_database.TrajectoryMetadata(
+            n_configurations=n_configs,
+            box_l=box_l,
+            sample_rate=sample_rate,
+            species_list=species_list,
+        )
+        self._batch_size = mdsuite.utils.meta_functions.optimize_batch_size(
+            filepath=self.file_path, number_of_configurations=n_configs
+        )
+
+        return self._mdata
```

Comment:
It feels like this should be a property. 
First because I like properties, but also because you use it internally as `self._mdata` and externally via `metadata = reader.get_metadata()` so why not combine both into a single `self.metadata` / `reader.metadata` and use 

```py
if self._metadata is None:
 self._metadata = self._get_metadata()
return self._metadata
```